Allow setting FQDN in hostname

javidiaz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature
/milestone 1.19

**What this PR does / why we need it**:
This PR enables a new feature that will allow users to set a pod’s hostname to its Fully Qualified Domain Name (FQDN). 

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes Kubernetes issue #91036


**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
Yes
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
setHostnameAsFQDN is a new field in PodSpec. When set to true, the fully 
qualified domain name (FQDN) of a Pod is set as hostname of its containers. 
In Linux containers, this means setting the FQDN in the hostname field of the 
kernel (the nodename field of struct utsname).  In Windows containers, this
means setting the this means setting the registry value of hostname for the registry key HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Services\\Tcpip\\Parameters to FQDN. 
If a pod does not have FQDN, this has no effect.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
- KEP: [enhancements/issues/1797](https://github.com/kubernetes/enhancements/issues/1797) [enhancements/pull/1792](https://github.com/kubernetes/enhancements/pull/1792)

- Usage: [kubernetes/website/pull/21210](https://github.com/kubernetes/website/pull/21210)

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

javidiaz committed:
Adding new PodSpec field called setHostnameAsFQDN and SetHostnameAsFQDN gate

These changes add a new field, called setHostnameAsFQDN, to the PodSpec. This
field is a bool that will be used to indicate whether we would like
FQDN be set as hostname or not.

This is PART1 of the changes to enable KEP #1797 and addresses #91036

javidiaz committed:
Adding autogenerated files

javidiaz committed:
Adding Kubelet changes to enable SetHostnameAsFQDN feature

These changes allow to set FQDN as hostname of pods for pods
that set the new PodSpec field setHostnameAsFQDN to true. The PodSpec
new field was added in related PR.

This is PART2 (last) of the changes to enable KEP #1797 and addresses #91036

javidiaz commented:
/sig node network

tedyu commented on pkg/kubelet/container/helpers.go:
```diff
@@ -52,6 +52,7 @@ type RuntimeHelper interface {
 	GetPodCgroupParent(pod *v1.Pod) string
 	GetPodDir(podUID types.UID) string
 	GeneratePodHostNameAndDomain(pod *v1.Pod) (hostname string, hostDomain string, err error)
+	GetHostNameForKernel(hostname string, hostDomainName string, hostnameFQDN bool) (string, error)
```
In other places, Hostname is used.
Why is HostName used here ?

javidiaz replied:
agree, i will change

tedyu commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -435,6 +435,21 @@ func (kl *Kubelet) GetPodCgroupParent(pod *v1.Pod) string {
+	if len(hostDomainName) > 0 && hostnameFQDN {
+		fqdnHostname := fmt.Sprintf("%s.%s", hostname, hostDomainName)
+		// FQDN has to be 63 chars to fit in hostname (specification is 64bytes which is 63 chars and the null terminating char).
+		const hostnameMaxLen = 63
```
This constant is already in kubelet_pods.go
Can you use the current const ?

javidiaz replied:
the other one is inside the function truncatePodHostnameIfNeeded, I will make it a global variable so we can reuse it.

tedyu commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -96,11 +96,14 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxConfig(pod *v1.Pod, attemp
+		if err != nil {
+			return nil, err
+		}
+		podSandboxConfig.Hostname, err = m.runtimeHelper.GetHostNameForKernel(hostname, hostDomain, pod.Spec.HostnameFQDN)
```
It seems podSandboxConfig.Hostname shouldn't be assigned if err is not nil

javidiaz replied:
agree, changing

tedyu commented on test/e2e/node/pod_hostnamefqdn.go:
```diff
@@ -0,0 +1,144 @@
+)
+
+func generatePodName(base string) string {
+	id, err := rand.Int(rand.Reader, big.NewInt(214748))
```
How is 214748 determined ?

javidiaz replied:
just some random numbers, I copied this function from somewhere and made the number smaller because my FQDN has to fit in 63 chars. I could not find any utility function to generate random pod name of fix size.

tedyu commented:
/ok-to-test

javidiaz commented on pkg/apis/core/types.go:
```diff
@@ -2693,6 +2693,12 @@ type PodSpec struct {
+	// In Windows containers, this means setting the COMPUTERNAME to FQDN.
+	// If a pod does not have FQDN, this has no effect.
+	// +optional
+	HostnameFQDN bool
```
Bringing conversation to this PR so we have all in one place

@tedyu:
Looking at existing bool fields, they start with a verb (such as Enable).

How about naming this field SetHostnameAsFQDN ?

@javidiaz:
I am open to change it. During the KEP review, I was suggested to use hostnameFQDN or fqdnHostname ( see kubernetes/enhancements#1792 (comment)). @thockin What do you think about changing the field as @tedyu suggest?

javidiaz replied:
I like SetHostnameAsFQDN, but I would like Tim to chime in here since he reviewed the KEP and I think he might do the API review.

thockin replied:
Fine

javidiaz replied:
thanks, changing to SetHostnameAsFQDN

javidiaz commented:
/assign @thockin

javidiaz commented:
/retest

javidiaz commented:
/label api-review
This is part of an approved KEP [enhancements/issues/1797](https://github.com/kubernetes/enhancements/issues/1797) [enhancements/pull/1792](https://github.com/kubernetes/enhancements/pull/1792)

bsdnet commented on pkg/kubelet/container/testing/fake_runtime_helper.go:
```diff
@@ -58,6 +59,14 @@ func (f *FakeRuntimeHelper) GeneratePodHostNameAndDomain(pod *v1.Pod) (string, s
 	return f.HostName, f.HostDomain, f.Err
 }
 
+func (f *FakeRuntimeHelper) GetHostNameForKernel(hostname string, hostDomainName string, hostnameFQDN bool) (string, error) {
```
Should this be GetNodenameForKernel since it bridges pod's hostname to kernel's nodename?

javidiaz replied:
sounds good.

bsdnet commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -435,6 +435,21 @@ func (kl *Kubelet) GetPodCgroupParent(pod *v1.Pod) string {
+	kernelhostname := hostname
+	if len(hostDomainName) > 0 && hostnameFQDN {
+		fqdnHostname := fmt.Sprintf("%s.%s", hostname, hostDomainName)
+		// FQDN has to be 63 chars to fit in hostname (specification is 64bytes which is 63 chars and the null terminating char).
```
Where is the spec, kernel source points to 64+1 or 8+1
https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/utsname.h#L5

javidiaz replied:
Good call, I think you are right. Each element of DNS has to be 63, but the utsname can be 64, so in my case the fqdn can be up to 64. I will update.

bsdnet commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -435,6 +435,21 @@ func (kl *Kubelet) GetPodCgroupParent(pod *v1.Pod) string {
+		// FQDN has to be 63 chars to fit in hostname (specification is 64bytes which is 63 chars and the null terminating char).
+		const hostnameMaxLen = 63
+		if len(fqdnHostname) > hostnameMaxLen {
+			return "", fmt.Errorf("Failed to set FQDN in hostname, Pod hostname %s is too long (%d characters is the limit)", fqdnHostname, hostnameMaxLen)
```
Failed to get FQDN from pod  hostname

javidiaz replied:
thanks, rephrasing

thockin replied:
Also say how many characters we have, eg "max is 64, requested is 72"

javidiaz replied:
ok, adding requested length too

bsdnet commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -435,6 +435,21 @@ func (kl *Kubelet) GetPodCgroupParent(pod *v1.Pod) string {
+func (kl *Kubelet) GetHostNameForKernel(hostname string, hostDomainName string, hostnameFQDN bool) (string, error) {
+	kernelhostname := hostname
+	if len(hostDomainName) > 0 && hostnameFQDN {
+		fqdnHostname := fmt.Sprintf("%s.%s", hostname, hostDomainName)
```
s/fqdnHostname/fqdn/g

looks we are form FQDN using pod's hostname and domain

javidiaz replied:
fixing

bsdnet commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -435,6 +435,21 @@ func (kl *Kubelet) GetPodCgroupParent(pod *v1.Pod) string {
 	return cgroupParent
 }
 
+// GetHostNameForKernel gets hostname value to set in the hostname field (the nodename field of struct utsname) of the pod.
```
GetNodenameForKernel is more explicit here.

javidiaz replied:
thanks, changing.

javidiaz commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3056,6 +3056,13 @@ type PodSpec struct {
+	// If a pod does not have FQDN, this has no effect.
+	// Default to false.
+	// +optional
+	HostnameFQDN bool `json:"hostnameFQDN,omitempty" protobuf:"varint,35,opt,name=hostnameFQDN"`
```
34 is in use https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L2880

javidiaz commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -435,6 +435,21 @@ func (kl *Kubelet) GetPodCgroupParent(pod *v1.Pod) string {
 
+// GetHostNameForKernel gets hostname value to set in the hostname field (the nodename field of struct utsname) of the pod.
+func (kl *Kubelet) GetHostNameForKernel(hostname string, hostDomainName string, hostnameFQDN bool) (string, error) {
+	kernelhostname := hostname
```
fixing, thanks

javidiaz commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -96,11 +96,14 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxConfig(pod *v1.Pod, attemp
 	if !kubecontainer.IsHostNetworkPod(pod) {
 		// TODO: Add domain support in new runtime interface
-		hostname, _, err := m.runtimeHelper.GeneratePodHostNameAndDomain(pod)
+		hostname, hostDomain, err := m.runtimeHelper.GeneratePodHostNameAndDomain(pod)
```
yeah, changing

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2693,6 +2693,12 @@ type PodSpec struct {
 	// If not specified, the pod will not have a domainname at all.
 	// +optional
 	Subdomain string
+	// If set to true, the fully qualified domain name (FQDN) of a Pod is set as hostname of its containers.
```
Reword for clarity, maybe :

If true the pod's hostname will be configured as the pod's FQDN, rather than the leaf name (the default).

javidiaz replied:
changing, thanks

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2693,6 +2693,12 @@ type PodSpec struct {
+	// In Windows containers, this means setting the COMPUTERNAME to FQDN.
+	// If a pod does not have FQDN, this has no effect.
+	// +optional
+	HostnameFQDN bool
```
There's a general convention of optionals always being pointers.  It is not 100% followed.  In this case it is kind of tiresome because it adds very little value to the system (since false is correct as the default).  @liggitt or @lavalamp - I feel like I am swimming upstream on this - are others enforcing this convention?

lavalamp replied:
Not making it a pointer means that users who construct their objects with the go typed objects cannot express the difference between "I want this off" and "I don't care". Doesn't matter for pod, but it could matter for the places where there are pod templates. SSA uses this bit to determine whether or not to give other users changing the field conflicts.

I don't think it's too likely that a completely separate actor would want to own this field vs other fields in pod templates (e.g., doesn't make sense in the way HPA -> replicas or VPA -> resource request). So it might not be super important.

Also users who directly construct the object in yaml/json do not have this problem.

javidiaz replied:
@thockin updated to use *bool

lavalamp replied:
Ah sorry for lack of clarity! My overall point was that it's probably OK to leave it a non pointer in this case. But of course I'm fine with it being a pointer.

thockin reviewed (commented):
I'll leave the kubelet parts for people more familiar with that code.

The last "review comments" commit really needs to be folded back into the "real" commits, or else squash the whole thing at the end?

javidiaz commented:
> I'll leave the kubelet parts for people more familiar with that code.
> 
> The last "review comments" commit really needs to be folded back into the "real" commits, or else squash the whole thing at the end?

@thockin do you mean I should be doing rebase and push -f to keep things align with the original 3 commits?

javidiaz commented:
/retest

javidiaz commented:
/test pull-kubernetes-e2e-kind-ipv6

javidiaz commented:
/retest

javidiaz commented:
@tedyu @bsdnet I think I addressed your comments, could you please double check and let me know if there is anything else you would like to be changed? If it is all good, could you add lgtm tag? 

@thockin I changed API field to be *bool, could you please have a look to see if api-review can be completed? Currently, you are the approver of this PR, are you ok with that or should we make someone from Sig-Node approver?

Thanks!

SergeyKanzhelev commented on pkg/kubelet/container/testing/fake_runtime_helper.go:
```diff
@@ -58,6 +59,14 @@ func (f *FakeRuntimeHelper) GeneratePodHostNameAndDomain(pod *v1.Pod) (string, s
+	if len(hostDomainName) > 0 && setHostnameAsFQDN != nil && *setHostnameAsFQDN {
+		return fmt.Sprintf("%s.%s", hostname, hostDomainName), f.Err
+	} else {
+		return fmt.Sprintf("%s", hostname), f.Err
```
nit: no need to format the string =)

javidiaz replied:
right, updated. thanks

SergeyKanzhelev commented on test/e2e/node/pod_hostnamefqdn.go:
```diff
@@ -0,0 +1,146 @@
+limitations under the License.
+*/
+
+/* This test check that SecurityContext parameters specified at the
```
Curious, is this a copy/paste?

javidiaz replied:
yes, i copy/paste from another e2e. I updated comment, thanks

SergeyKanzhelev commented on test/e2e/node/pod_hostnamefqdn.go:
```diff
@@ -0,0 +1,146 @@
+	})
+
+	/*
+		   Release : v1.19
+		   Testname: Create Pod with FQDN, setHostnameAsFQDN field set to true.
+		   Description: A Pod that defines the subdomain field in it spec has FQDN. When setHostnameAsFQDN: true, the
+				hostname is set to be the FQDN. In this case, both commands hostname and hostname -f return
+				the FQDN of the Pod.
```
```suggestion
	   Release : v1.19
	   Testname: Create Pod with FQDN, setHostnameAsFQDN field set to true.
	   Description: A Pod that defines the subdomain field in it spec has FQDN. When setHostnameAsFQDN: true, the
				hostname is set to be the FQDN. In this case, both commands hostname and hostname -f return
				the FQDN of the Pod.
```

javidiaz replied:
yes, I knew about this one and i could not get gomft happy, i think i finally figured out.

javidiaz commented:
/test pull-kubernetes-integration

javidiaz commented:
I tested in Windows and the feature does not cause any issues. I am updating comments as it seems that COMPUTERNAME environment variable is not set as I thought. In turn, it is some property in the Windows registry, which seems to be used by the hostname command.

> Windows PowerShell
> Copyright (C) Microsoft Corporation. All rights reserved.
> PS C:\> hostname
> win-webserverfqdn-86484565d6-8fr88.sub.default.svc.cluster.local
> PS C:\> $env:computername 
> WIN-WEBSERVERFQ 
> PS C:\>

The registry value is

> 
> PS HKLM:\SYSTEM\CurrentControlSet\Services\Tcpip> ls    Hive: HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\TcpipName                           Property
> ----                           --------
> ....
> Parameters                     NV HostName              : win-webserverfqdn-86484565d6-8fr88.sub.default.svc.cluster.local
>                                Hostname                 : win-webserverfqdn-86484565d6-8fr88.sub.default.svc.cluster.local
> ...

javidiaz commented:
/retest

javidiaz commented:
/retest

javidiaz committed:
Adding Bazel deps

javidiaz commented:
/retest

javidiaz commented:
/retest

thockin commented:
I'm LGTM on this, but it would be nice for someone in sig-node to weight in.

@dchen1107 @derekwaynecarr @mrunalp @sjenning

thockin commented:
/lgtm

dashpole commented:
kubelet changes and node_e2e test
/lgtm

dchen1107 commented:
/lgtm
/approve

thockin commented:
/approve
